File: c/Salsa20/salsa20.c

#include "salsa20.h"
#include <string.h>

#define ROTL(a,b) (((a) << (b)) | ((a) >> (32 - (b))))

#define QR(a,b,c,d) \
	b ^= ROTL(a + d, 7); \
	c ^= ROTL(b + a, 9); \
	d ^= ROTL(c + b, 13); \
	a ^= ROTL(d + c, 18);

static const char constant[16] = "expand 32-byte k";

static uint32_t load32_le(const uint8_t *src) {
    return ((uint32_t)src[0]) |
           ((uint32_t)src[1] << 8) |
           ((uint32_t)src[2] << 16) |
           ((uint32_t)src[3] << 24);
}

static void store32_le(uint8_t *dst, uint32_t val) {
    dst[0] = (uint8_t)(val);
    dst[1] = (uint8_t)(val >> 8);
    dst[2] = (uint8_t)(val >> 16);
    dst[3] = (uint8_t)(val >> 24);
}

void salsa20_init(salsa20_context *ctx, const uint8_t key[32], const uint8_t nonce[8]) {
    // Constants "expand 32-byte k"
    // sigma = "expand 32-byte k"
    ctx->state[0]  = load32_le((const uint8_t*)constant);
    ctx->state[5]  = load32_le((const uint8_t*)constant + 4);
    ctx->state[10] = load32_le((const uint8_t*)constant + 8);
    ctx->state[15] = load32_le((const uint8_t*)constant + 12);

    // Key
    ctx->state[1] = load32_le(key);
    ctx->state[2] = load32_le(key + 4);
    ctx->state[3] = load32_le(key + 8);
    ctx->state[4] = load32_le(key + 12);
    ctx->state[11] = load32_le(key + 16);
    ctx->state[12] = load32_le(key + 20);
    ctx->state[13] = load32_le(key + 24);
    ctx->state[14] = load32_le(key + 28);

    // Nonce
    ctx->state[6] = load32_le(nonce);
    ctx->state[7] = load32_le(nonce + 4);

    // Block counter (64-bit), starts at 0
    ctx->state[8] = 0;
    ctx->state[9] = 0;
}

static void salsa20_block(uint32_t state[16], uint8_t output[64]) {
    uint32_t x[16];
    int i;

    memcpy(x, state, 64);

    for (i = 0; i < 10; i++) {
        // Column rounds
        QR(x[ 0], x[ 4], x[ 8], x[12]);
        QR(x[ 5], x[ 9], x[13], x[ 1]);
        QR(x[10], x[14], x[ 2], x[ 6]);
        QR(x[15], x[ 3], x[ 7], x[11]);

        // Row rounds
        QR(x[ 0], x[ 1], x[ 2], x[ 3]);
        QR(x[ 5], x[ 6], x[ 7], x[ 4]);
        QR(x[10], x[11], x[ 8], x[ 9]);
        QR(x[15], x[12], x[13], x[14]);
    }

    for (i = 0; i < 16; i++) {
        x[i] += state[i];
        store32_le(output + i * 4, x[i]);
    }
}

void salsa20_crypt(salsa20_context *ctx, const uint8_t *input, uint8_t *output, size_t len) {
    uint8_t block[64];
    size_t i = 0;

    while (len > 0) {
        salsa20_block(ctx->state, block);

        // Increment 64-bit counter (state[8] and state[9])
        ctx->state[8]++;
        if (ctx->state[8] == 0) {
            ctx->state[9]++;
        }

        size_t n = (len < 64) ? len : 64;
        for (size_t j = 0; j < n; j++) {
            output[i + j] = input[i + j] ^ block[j];
        }

        i += n;
        len -= n;
    }
}

File: js/Salsa20/Salsa20.js

// Salsa20.js
// Ported from c/Salsa20/salsa20.c

class Salsa20 {
    constructor(key, nonce) {
        this.state = new Uint32Array(16);
        this.init(key, nonce);
    }

    static ROTL(a, b) {
        return ((a << b) | (a >>> (32 - b))) >>> 0;
    }

    static load32_le(src, offset) {
        return (src[offset] |
               (src[offset + 1] << 8) |
               (src[offset + 2] << 16) |
               (src[offset + 3] << 24)) >>> 0;
    }

    static store32_le(dst, offset, val) {
        dst[offset] = val & 0xFF;
        dst[offset + 1] = (val >>> 8) & 0xFF;
        dst[offset + 2] = (val >>> 16) & 0xFF;
        dst[offset + 3] = (val >>> 24) & 0xFF;
    }

    init(key, nonce) {
        // Constants "expand 32-byte k"
        const constant = [0x61707865, 0x3320646e, 0x79622d32, 0x6b206574];
        
        this.state[0] = constant[0];
        this.state[5] = constant[1];
        this.state[10] = constant[2];
        this.state[15] = constant[3];

        // Key
        this.state[1] = Salsa20.load32_le(key, 0);
        this.state[2] = Salsa20.load32_le(key, 4);
        this.state[3] = Salsa20.load32_le(key, 8);
        this.state[4] = Salsa20.load32_le(key, 12);
        this.state[11] = Salsa20.load32_le(key, 16);
        this.state[12] = Salsa20.load32_le(key, 20);
        this.state[13] = Salsa20.load32_le(key, 24);
        this.state[14] = Salsa20.load32_le(key, 28);

        // Nonce
        this.state[6] = Salsa20.load32_le(nonce, 0);
        this.state[7] = Salsa20.load32_le(nonce, 4);

        // Block counter (64-bit), starts at 0
        this.state[8] = 0;
        this.state[9] = 0;
    }

    block(output) {
        const x = new Uint32Array(this.state);

        for (let i = 0; i < 10; i++) {
            // Column rounds
            this.QR(x, 0, 4, 8, 12);
            this.QR(x, 5, 9, 13, 1);
            this.QR(x, 10, 14, 2, 6);
            this.QR(x, 15, 3, 7, 11);

            // Row rounds
            this.QR(x, 0, 1, 2, 3);
            this.QR(x, 5, 6, 7, 4);
            this.QR(x, 10, 11, 8, 9);
            this.QR(x, 15, 12, 13, 14);
        }

        for (let i = 0; i < 16; i++) {
            x[i] = (x[i] + this.state[i]) >>> 0;
            Salsa20.store32_le(output, i * 4, x[i]);
        }
    }

    QR(x, a, b, c, d) {
        x[b] ^= Salsa20.ROTL((x[a] + x[d]) >>> 0, 7);
        x[c] ^= Salsa20.ROTL((x[b] + x[a]) >>> 0, 9);
        x[d] ^= Salsa20.ROTL((x[c] + x[b]) >>> 0, 13);
        x[a] ^= Salsa20.ROTL((x[d] + x[c]) >>> 0, 18);
    }

    crypt(input) {
        let inputBytes;
        if (typeof input === 'string') {
            inputBytes = new TextEncoder().encode(input);
        } else {
            inputBytes = input;
        }

        const len = inputBytes.length;
        const output = new Uint8Array(len);
        const blockBuffer = new Uint8Array(64);
        let i = 0;
        let remaining = len;

        while (remaining > 0) {
            this.block(blockBuffer);

            // Increment 64-bit counter
            this.state[8] = (this.state[8] + 1) >>> 0;
            if (this.state[8] === 0) {
                this.state[9] = (this.state[9] + 1) >>> 0;
            }

            const n = (remaining < 64) ? remaining : 64;
            for (let j = 0; j < n; j++) {
                output[i + j] = inputBytes[i + j] ^ blockBuffer[j];
            }

            i += n;
            remaining -= n;
        }

        return output;
    }
}

module.exports = Salsa20;
